### PR TITLE
[MISC] adjusting seqan3_version semantics

### PR DESCRIPTION
**Pull request by SGSSGene**
- Adjusting seqan3_version to be the constexpr version of SEQAN_VERSION which is
  an integer.
- Added SEQAN3_VERSION_CSTRING which is a null terminated string of the version.
- Added seqan3_version_cstring which is a constexpr version of
  SEQAN3_VERSION_CSTRING.
- Adjusted all files to use seqan3_version_cstring instead of seqan3_version.

### CODE REVIEW

**Comment by SGSSGene** [comment]
Corresponding Issue can be found here: https://github.com/seqan/product_backlog/issues/277

File: include/seqan3/version.hpp [hunk 1]
```diff
@@ -28,6 +28,24 @@
                      + SEQAN3_VERSION_MINOR * 100 \
                      + SEQAN3_VERSION_PATCH)
 
+/*!\brief Converts a number to a string.
+ * Preprocessor needs this indirection to
```
Comment (MitraDarja) [inline include/seqan3/version.hpp:32]: I think, you can put these on two lines, instead of three.

**Review by MitraDarja — CHANGES_REQUESTED** [review]
LGTM, just a really small thing. :)

**Comment by MitraDarja** [comment]
And I think, this should not be CHANGE but MISC? http://docs.seqan.de/seqan/3-master-user/about_contributing.html

File: include/seqan3/version.hpp (hunk 1 above)
Comment (SGSSGene) [reply include/seqan3/version.hpp]: Yes, I should do that and will! :-)

**Review by MitraDarja — APPROVED** [review]
Thanks. :)

File: include/seqan3/version.hpp [hunk 2]
```diff
@@ -28,6 +28,23 @@
...
+    SEQAN3_VERSION_CSTRING_HELPER_STR(MAJOR) "."\
+    SEQAN3_VERSION_CSTRING_HELPER_STR(MINOR) "."\
+    SEQAN3_VERSION_CSTRING_HELPER_STR(PATCH)
+
+//!\brief The full version as null terminated string.
+#define SEQAN3_VERSION_CSTRING \
+    SEQAN3_VERSION_CSTRING_HELPER_FUNC(SEQAN3_VERSION_MAJOR, \
+                                       SEQAN3_VERSION_MINOR, \
+                                       SEQAN3_VERSION_PATCH)
+
```
Comment (marehr) [inline include/seqan3/version.hpp:47]:
```suggestion
//!\cond DEV
/*!\brief Converts a number to a string. Preprocessor needs this indirection to
 * properly expand the values to strings.
 */
#define SEQAN3_VERSION_CSTRING_HELPER_STR(x) #x

//!\brief Converts version numbers to string.
#define SEQAN3_VERSION_CSTRING_HELPER_FUNC(MAJOR, MINOR, PATCH) \
    SEQAN3_VERSION_CSTRING_HELPER_STR(MAJOR) "."\
    SEQAN3_VERSION_CSTRING_HELPER_STR(MINOR) "."\
    SEQAN3_VERSION_CSTRING_HELPER_STR(PATCH)
//!\endcond

//!\brief The full version as null terminated string.
#define SEQAN3_VERSION_CSTRING \
    SEQAN3_VERSION_CSTRING_HELPER_FUNC(SEQAN3_VERSION_MAJOR, \
                                       SEQAN3_VERSION_MINOR, \
                                       SEQAN3_VERSION_PATCH)

#undef SEQAN3_VERSION_CSTRING_HELPER_STR
#undef SEQAN3_VERSION_CSTRING_HELPER_FUNC
```

Comment (marehr) [reply include/seqan3/version.hpp:47]: We shouldn't leak these helpers

File: test/unit/argument_parser/detail/format_help_test.cpp [hunk 3]
```diff
@@ -43,7 +43,7 @@ std::string const basic_options_str = "OPTIONS\n"
 std::string const basic_version_str = "VERSION\n"
                                       "    Last update:\n"
                                       "    test_parser version:\n"
-                                      "    SeqAn version: " + seqan3::seqan3_version + "\n";
+                                      "    SeqAn version: " + std::string(seqan3::seqan3_version_cstring) + "\n";
```
Comment (marehr) [inline test/unit/argument_parser/detail/format_help_test.cpp:46]:
```suggestion
                                      "    SeqAn version: " + std::string{seqan3::seqan3_version_cstring} + "\n";
```

File: test/unit/argument_parser/detail/format_html_test.cpp [hunk 4]
```diff
@@ -53,7 +53,7 @@ TEST(html_format, empty_information)
                            "<br>\n"
                            "<strong>empty_options version: </strong>\n"
                            "<br>\n"
-                           "<strong>SeqAn version: </strong>" + seqan3::seqan3_version + "\n"
+                           "<strong>SeqAn version: </strong>" + std::string(seqan3::seqan3_version_cstring) + "\n"
```
Comment (marehr) [inline test/unit/argument_parser/detail/format_html_test.cpp:56]:
```suggestion
                           "<strong>SeqAn version: </strong>" + std::string{seqan3::seqan3_version_cstring} + "\n"
```

File: test/unit/argument_parser/detail/format_html_test.cpp [hunk 5]
```diff
@@ -167,7 +167,7 @@ TEST(html_format, full_information_information)
                           "<br>\n"
                           "<strong>program_full_options version: </strong>\n"
                           "<br>\n"
-                          "<strong>SeqAn version: </strong>" + seqan3::seqan3_version + "\n"
+                          "<strong>SeqAn version: </strong>" + std::string(seqan3::seqan3_version_cstring) + "\n"
```
Comment (marehr) [inline test/unit/argument_parser/detail/format_html_test.cpp:170]:
```suggestion
                          "<strong>SeqAn version: </strong>" + std::string{seqan3::seqan3_version_cstring} + "\n"
```

File: test/unit/argument_parser/detail/version_check_test.hpp [hunk 6]
```diff
@@ -390,7 +390,7 @@ TEST_F(version_check, greater_app_version)
     const char * argv[3] = {app_name.c_str(), OPTION_VERSION_CHECK, OPTION_ON};
 
     // create version file with equal seqan version and a smaller app version than the current
-    ASSERT_TRUE(create_file(app_version_filename(), std::string{"1.5.9\n" + seqan3::seqan3_version}));
+    ASSERT_TRUE(create_file(app_version_filename(), std::string{"1.5.9\n" + std::string(seqan3::seqan3_version_cstring)}));
```
Comment (marehr) [inline test/unit/argument_parser/detail/version_check_test.hpp:393]:
```suggestion
    ASSERT_TRUE(create_file(app_version_filename(), std::string{"1.5.9\n"} + seqan3::seqan3_version_cstring));
```

File: test/unit/argument_parser/detail/version_check_test.hpp [hunk 7]
```diff
@@ -411,7 +411,7 @@ TEST_F(version_check, unregistered_app)
     const char * argv[3] = {app_name.c_str(), OPTION_VERSION_CHECK, OPTION_ON};
 
     // create version file with equal seqan version and a smaller app version than the current
-    ASSERT_TRUE(create_file(app_version_filename(), std::string{ "UNREGISTERED_APP\n" + seqan3::seqan3_version}));
+    ASSERT_TRUE(create_file(app_version_filename(), std::string{ "UNREGISTERED_APP\n" + std::string(seqan3::seqan3_version_cstring)}));
```
Comment (marehr) [inline test/unit/argument_parser/detail/version_check_test.hpp:414]:
```suggestion
    ASSERT_TRUE(create_file(app_version_filename(), std::string{"UNREGISTERED_APP\n"} + seqan3::seqan3_version_cstring));
```

File: test/unit/argument_parser/detail/version_check_test.hpp [hunk 8]
```diff
@@ -435,7 +435,7 @@ TEST_F(version_check, smaller_app_version)
     const char * argv[3] = {app_name.c_str(), OPTION_VERSION_CHECK, OPTION_ON};
 
     // create version file with equal seqan version and a greater app version than the current
-    ASSERT_TRUE(create_file(app_version_filename(), std::string{"20.5.9\n" + seqan3::seqan3_version}));
+    ASSERT_TRUE(create_file(app_version_filename(), std::string{"20.5.9\n" + std::string(seqan3::seqan3_version_cstring)}));
```
Comment (marehr) [inline test/unit/argument_parser/detail/version_check_test.hpp:438]:
```suggestion
    ASSERT_TRUE(create_file(app_version_filename(), std::string{"20.5.9\n"} + seqan3::seqan3_version_cstring));
```

File: test/unit/argument_parser/detail/version_check_test.hpp [hunk 9]
```diff
@@ -463,7 +463,7 @@ TEST_F(version_check, smaller_app_version_custom_url)
...
```
Comment (marehr) [inline test/unit/argument_parser/detail/version_check_test.hpp:466]:
```suggestion
    ASSERT_TRUE(create_file(app_version_filename(), std::string{"20.5.9\n"} + seqan3::seqan3_version_cstring));
```

File: test/unit/argument_parser/format_parse_validators_test.cpp [hunk 10]
```diff
@@ -47,7 +47,7 @@ std::string const basic_options_str = "OPTIONS\n"
 std::string const basic_version_str = "VERSION\n"
                                       "    Last update:\n"
                                       "    test_parser version:\n"
-                                      "    SeqAn version: " + seqan3::seqan3_version + "\n";
+                                      "    SeqAn version: " + std::string(seqan3::seqan3_version_cstring) + "\n";
```
Comment (marehr) [inline test/unit/argument_parser/format_parse_validators_test.cpp:50]:
```suggestion
                                      "    SeqAn version: " + std::string{seqan3::seqan3_version_cstring} + "\n";
```

**Review by marehr — COMMENTED** [review]
Thank you for the quick PR.

Just one site note, we use brace-intialisation where possible.

File: include/seqan3/version.hpp [hunk 11]
```diff
@@ -28,6 +28,27 @@
...
+//!\endcond
+
+//!\brief The full version as null terminated string.
+#define SEQAN3_VERSION_CSTRING \
+    SEQAN3_VERSION_CSTRING_HELPER_FUNC(SEQAN3_VERSION_MAJOR, \
+                                       SEQAN3_VERSION_MINOR, \
+                                       SEQAN3_VERSION_PATCH)
+
+#undef SEQAN3_VERSION_CSTRING_HELPER_STR
+#undef SEQAN3_VERSION_CSTRING_HELPER_FUNC
```
Comment (marehr) [inline include/seqan3/version.hpp:51]:
```suggestion
#undef SEQAN3_VERSION_CSTRING_HELPER_FUNC

```

**Comment by marehr** [comment]
What is open: remove all "heavy" includes from `version.hpp`

File: include/seqan3/version.hpp [hunk 12]
```diff
@@ -28,6 +28,28 @@
...
```
Comment (eseiler) [inline include/seqan3/version.hpp:51]:
```suggestion
```

File: include/seqan3/version.hpp [hunk 13]
```diff
@@ -38,9 +60,10 @@ constexpr uint8_t seqan3_version_minor = SEQAN3_VERSION_MINOR;
...
-std::string const seqan3_version = std::to_string(seqan3_version_major) + "." +
-                                   std::to_string(seqan3_version_minor) + "." +
-                                   std::to_string(seqan3_version_patch);
+//!\brief The full version as `std::size_t`.
+constexpr std::size_t seqan3_version = SEQAN3_VERSION;
+
+//!\brief The full version as null terminated string.
+constexpr char const* seqan3_version_cstring = SEQAN3_VERSION_CSTRING;
 
 } // namespace seqan3
```
Comment (eseiler) [inline include/seqan3/version.hpp:69]:
```suggestion
} // namespace seqan3

#undef SEQAN3_VERSION_CSTRING_HELPER_STR
#undef SEQAN3_VERSION_CSTRING_HELPER_FUNC
```

**Review by eseiler — COMMENTED** [review]
macro was undefined before it was used

File: include/seqan3/version.hpp (hunk 12 above)
Comment (marehr) [reply include/seqan3/version.hpp]: thx :)

**Review by marehr — APPROVED** [review]

**Pull request merged**